File: truevfs-kernel/truevfs-kernel-impl/src/main/scala/net/java/truevfs/kernel/impl/ReadOnlyArchiveFileSystem.scala
/*
 * Copyright (C) 2005-2015 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.kernel.impl

import javax.annotation.concurrent._
import net.java.truecommons.cio.Entry.Access._
import net.java.truecommons.cio.Entry._
import net.java.truecommons.cio._
import net.java.truecommons.shed._
import net.java.truevfs.kernel.impl.ReadOnlyArchiveFileSystem._
import net.java.truevfs.kernel.spec._

import scala.Option

/** A read-only virtual file system for archive entries.
  *
  * All modifying methods throw a
  * [[net.java.truevfs.kernel.spec.FsReadOnlyFileSystemException]].
  *
  * @tparam E The type of the archive entries.
  * @author Christian Schlichtherle
  */
@NotThreadSafe
private final class ReadOnlyArchiveFileSystem[E <: FsArchiveEntry]
(model: ArchiveModel[E], archive: Container[E], rootTemplate: Entry, cause: Throwable)
  extends ArchiveFileSystem(model, archive, rootTemplate) {

  override def checkAccess(options: AccessOptions, name: FsNodeName, types: BitField[Access]): Unit = {
    if (!types.isEmpty && READ_ONLY != types) {
      throw new FsReadOnlyFileSystemException(mountPoint, cause)
    }
    super.checkAccess(options, name, types)
  }

  override def setReadOnly(options: AccessOptions, name: FsNodeName): Unit = {
  }

  override def setTime(options: AccessOptions, name: FsNodeName, times: Map[Access, Long]): Boolean = {
    throw new FsReadOnlyFileSystemException(mountPoint, cause)
  }

  override def setTime(options: AccessOptions, name: FsNodeName, types: BitField[Access], value: Long): Boolean = {
    throw new FsReadOnlyFileSystemException(mountPoint, cause)
  }

  override def make(options: AccessOptions, name: FsNodeName, tµpe: Type, template: Option[Entry]): Make = {
    throw new FsReadOnlyFileSystemException(mountPoint, cause)
  }

  override def unlink(options: AccessOptions, name: FsNodeName): Unit = {
    throw new FsReadOnlyFileSystemException(mountPoint, cause)
  }
}

private object ReadOnlyArchiveFileSystem {

  private val READ_ONLY = BitField.of(READ)
}

File: truevfs-kernel/truevfs-kernel-impl/src/main/java/net/java/truevfs/kernel/impl/ReadOnlyArchiveFileSystem.java
/*
 * Copyright (C) 2005-2015 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.kernel.impl;

import net.java.truecommons.cio.Container;
import net.java.truecommons.cio.Entry;
import net.java.truecommons.shed.BitField;
import net.java.truevfs.kernel.spec.FsAccessOption;
import net.java.truevfs.kernel.spec.FsArchiveEntry;
import net.java.truevfs.kernel.spec.FsNodeName;
import net.java.truevfs.kernel.spec.FsReadOnlyFileSystemException;

import javax.annotation.concurrent.NotThreadSafe;
import java.io.IOException;
import java.util.Map;
import java.util.Optional;
import java.util.function.Supplier;

import static net.java.truecommons.cio.Entry.Access.READ;

/**
 * A read-only virtual file system for archive entries.
 * <p>
 * All modifying methods throw a {@link net.java.truevfs.kernel.spec.FsReadOnlyFileSystemException}.
 *
 * @param <E> The type of the archive entries.
 * @author Christian Schlichtherle
 */
@NotThreadSafe
final class ReadOnlyArchiveFileSystem<E extends FsArchiveEntry> extends ArchiveFileSystem<E> {

    private static final BitField<Entry.Access> READ_ONLY = BitField.of(READ);

    private final Supplier<? extends Throwable> cause;

    ReadOnlyArchiveFileSystem(
            final ArchiveModel<E> model,
            final Container<E> archive,
            final Entry rootTemplate,
            final Supplier<? extends Throwable> cause
    ) {
        super(model, archive, rootTemplate);
        this.cause = cause;
    }

    @Override
    void checkAccess(
            final BitField<FsAccessOption> options,
            final FsNodeName name,
            final BitField<Entry.Access> types
    ) throws IOException {
        if (!types.isEmpty() && READ_ONLY != types) {
            throw newFsReadOnlyFileSystemException();
        }
        super.checkAccess(options, name, types);
    }

    @Override
    void setReadOnly(BitField<FsAccessOption> options, FsNodeName name) throws IOException {
    }

    @Override
    boolean setTime(BitField<FsAccessOption> options, FsNodeName name, Map<Entry.Access, Long> times) throws IOException {
        throw newFsReadOnlyFileSystemException();
    }

    @Override
    boolean setTime(BitField<FsAccessOption> options, FsNodeName name, BitField<Entry.Access> types, long value) throws IOException {
        throw newFsReadOnlyFileSystemException();
    }

    @Override
    ArchiveFileSystem<E>.Make make(BitField<FsAccessOption> options, FsNodeName name, Entry.Type type, Optional<Entry> template) throws IOException {
        throw newFsReadOnlyFileSystemException();
    }

    @Override
    void unlink(BitField<FsAccessOption> options, FsNodeName name) throws IOException {
        throw newFsReadOnlyFileSystemException();
    }

    private FsReadOnlyFileSystemException newFsReadOnlyFileSystemException() {
        return new FsReadOnlyFileSystemException(getMountPoint(), cause.get());
    }
}
